docs/cpu_templates/cpu-templates.md

```diff
@@ -0,0 +1,168 @@
+
+- CPUID (x86_64 only)
+- MSRs (Model Specific Registers, x86_64 only)
+- ARM registers (aarch64 only)
+
+A combination of the changes to the entities above is called a CPU template.
+
+The functionality can be used when a user wants to mask a feature from
+the guest. One of real world use cases for this is representing a heterogeneous
+fleet (a fleet consisting of multiple CPU models) as a homogeneous fleet,
+so the guests will experience a consistent feature set supported by the host.
+
+Note: representing one CPU vendor as another CPU vendor is not supported.
+
+Firecracker supports two types of CPU templates:
+
+- Static CPU templates - a set of preconfigured CPU templates for users
+  to choose from;
+- Custom CPU templates - users can create their own CPU templates.
+
+## Static CPU templates
+
+At the moment the following set of static CPU templates is supported:
+
+| CPU template | CPU vendor | CPU model             |
+|--------------|------------|-----------------------|
+| C3           | Intel      | any                   |
+| T2           | Intel      | any                   |
+| T2A          | AMD        | Milan                 |
+| T2CL         | Intel      | Cascade Lake or newer |
+| T2S          | Intel      | any                   |
+| V1N1         | ARM        | Neoverse V1           |
+
+T2 and C3 templates are mapped as close as possible to AWS T2 and C3 instances
+in terms f CPU features.
+
+The T2S template is designed to allow migrating [snapshots](../snapshotting/versioning.md)
+between hosts with Intel Skylake and Intel Cascade Lake securely by further
+restricting CPU features for the guest, however this comes with a performance
+penalty. Users are encouraged to carry out a performance assessment if they wish
+to use the T2S template.
+
+The T2CL template is mapped to be close to Intel Cascade Lake.
+It is not safe to use it on Intel CPUs older than Cascade Lake (such as Skylake).
+
+The only AMD template is T2A. It is considered safe to be used with AMD Milan.
+
+Intel T2CL and AMD T2A templates together aim to provide instruction set feature
+parity between CPUs running them, so they can form a heterogeneous fleet
+exposing the same instruction sets to the application.
+
+The V1N1 template is designed to represent ARM Neoverse V1 as ARM Neoverse N1.
+
+### Configuring static CPU templates
+
+Configuration of a static CPU template is performed via `/machine-config` API endpoint:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i  \
+  -X PUT 'http://localhost/machine-config' \
+  -H 'Accept: application/json'            \
+  -H 'Content-Type: application/json'      \
+  -d '{
+           "vcpu_count": 2,
+           "mem_size_mib": 1024,
+           "cpu_template": "T2CL"
+  }'
+```
+
+## Custom CPU templates
+
+Users can create their own CPU templates by creating a json file containing
+modifiers for CPUID, MSRs or ARM registers.
+
+Note: Creating custom CPU templates requires expert knowledge of
```

Comment:
This can be 
"> \*\*Note\*\*" (without "")
to look like:
> **Note**

(Also works with warnings)
> **Warning**

Reply:
Also we need to be explicit about ARM cpu templates and that users will need patched kernel to make templates do anything useful.

Reply:
nice!

Reply:
> Also we need to be explicit about ARM cpu templates and that users will need patched kernel to make templates do anything useful.

Yes, that's a good point. Let me think how we can express that.

Reply:
added.